collect_app/src/main/java/org/odk/collect/android/http/HttpClientConnection.java

```diff
@@ -0,0 +1,708 @@
+
+        Header[] fields = response.getAllHeaders();
+
+        if (fields != null && fields.length >= 1) {
+            for (Header h : fields) {
+                responseHeaders.put(h.getName(), h.getValue());
+            }
+        }
+
+        return new HttpGetResult(downloadStream, responseHeaders, hash, statusCode);
+    }
+
+
+    @Override
+    public HttpHeadResult head(@NonNull URI uri, @Nullable HttpCredentialsInterface credentials) throws Exception {
+        addCredentialsForHost(uri, credentials);
+        getCookieStore().clear();
+
+        HttpContext localContext = getHttpContext();
+        HttpClient httpclient = createHttpClient(CONNECTION_TIMEOUT);
+        HttpHead httpHead = createOpenRosaHttpHead(uri);
+        Map<String, String> responseHeaders = new HashMap<>();
+
+        // if https then enable preemptive basic auth...
+        if (uri.getScheme() != null && uri.getScheme().equals("https")) {
+            enablePreemptiveBasicAuth(localContext, uri.getHost());
+        }
+
+        final HttpResponse response;
+        int statusCode;
+
+        try {
+            Timber.i("Issuing HEAD request to: %s", uri.toString());
+
+            response = httpclient.execute(httpHead, localContext);
+            statusCode = response.getStatusLine().getStatusCode();
+            if (statusCode == HttpStatus.SC_UNAUTHORIZED) {
+                getCookieStore().clear();
+
+            } else if (statusCode == HttpStatus.SC_NO_CONTENT) {
+                for (Header head : response.getAllHeaders()) {
+                    responseHeaders.put(head.getName(), head.getValue());
+                }
+            }
+
+            discardEntityBytes(response);
+
+        } catch (ClientProtocolException | ConnectTimeoutException | UnknownHostException | SocketTimeoutException | NoHttpResponseException | SocketException e) {
+            String errorMessage;
+
+            if (e instanceof ClientProtocolException) {
+                errorMessage = "Client Protocol Exception";
+            } else if (e instanceof ConnectTimeoutException) {
+                errorMessage = "Connection Timeout";
+            } else if (e instanceof UnknownHostException) {
+                errorMessage = e.toString() + " :: Network Connection Failed";
+            } else if (e instanceof SocketTimeoutException) {
+                errorMessage = "Connection Timeout";
+            } else {
+                errorMessage = "Network Connection Refused";
+            }
+
+            throw new Exception(errorMessage);
+
+        } catch (Exception e) {
+            String msg = e.getMessage();
+            if (msg == null) {
+                msg = e.toString();
+            }
+
+            throw new Exception("Generic Exception: " + msg);
+        }
+
+        return new HttpHeadResult(statusCode, responseHeaders);
+    }
+
+    @Override
+    public ResponseMessageParser uploadSubmissionFile(@NonNull List<File> fileList,
```

Comment:
Should this have `@NonNull` as well? Since the API design, it's following stipulates that?